cpp/src/arrow/compute/kernels/row_encoder_internal.h

```diff
@@ -287,18 +277,188 @@ void ProcessListScalar(const ScalarType& scalar, ValidFunc&& valid_func,
     int64_t list_length = scalar_data->length;
     const uint8_t* scalar_validity_buf = scalar_data->buffers[0]->data() == NULLPTR ?
       NULLPTR : scalar_data->buffers[0]->data();
-    const uint8_t* scalar_data_buf = scalar_data->buffers[1]->data();
+    const uint8_t* scalar_data_buf_1 = scalar_data->buffers[1]->data() == NULLPTR ?
+      NULLPTR: scalar_data->buffers[1]->data();
+    const uint8_t* scalar_data_buf_2 =
+      scalar_data->buffers.size() < 3 || scalar_data->buffers[2]->data() == NULLPTR ?
+      NULLPTR : scalar_data->buffers[2]->data();
 
-    valid_func(list_length, scalar_validity_buf, scalar_data_buf);
+    valid_func(list_length, scalar_validity_buf, scalar_data_buf_1, scalar_data_buf_2);
   } else { // data.scalar->is_valid == false
     null_func();
   }
 }
 
+// A template to know how much more to allocate for the encoding of a list type for a
+// batch.
+// get_list_length must take
+// - the starting offset of the list
+// - the ending offset of the list
+// - the start of the validity buffer of the nested-type
+// - the start of the nested-type's buffers[1]
+// - the start of the nested-type's buffers[2] or NULLPTR if it does not exist
+// - returns: the amount of bytes to allocate for the list; get_list_length(0, 0, NULLPTR,
+//   NULLPTR, NULLPTR) must return the amount of space to allocate for a null list
+template<typename T, typename LengthFunc>
+void AddListLength(const ExecValue& data, int64_t batch_length, int32_t* lengths,
+                  LengthFunc&& get_list_length) {
+  using Offset = typename T::offset_type;
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+
+  const int32_t null_list_length = get_list_length(0, 0, NULLPTR, NULLPTR, NULLPTR);
+  if (data.is_array()) {
+    const ArraySpan& arr = data.array;
+    const uint8_t* child_validity_buffer = arr.child_data[0].GetValues<uint8_t>(0);
+    const uint8_t* child_data_buffer_1 = arr.child_data[0].GetValues<uint8_t>(1);
+    const uint8_t* child_data_buffer_2 = arr.child_data[0].GetValues<uint8_t>(2);
```

Comment:
Do we need to check if the child data has less than 3 buffers?

Reply:
child data is an `ArraySpan`, which stores buffers as
```
  BufferSpan buffers[3];
```

It's pretty troll as this is not the case for `Scalars` with buffers, which is stored in a `vector` of buffers...